sklearn/metrics/pairwise.py

```diff
@@ -224,41 +234,156 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+            distances = _euclidean_distances_upcast_fast(X, XX, Y, YY)
 
-        if YY.shape != (1, Y.shape[0]):
-            raise ValueError(
-                "Incompatible dimensions for Y and Y_norm_squared")
+        # if dtype is already float64, no need to chunk and upcast
+        else:
+            if X is Y and not issparse(X):
+                # In this case the distance matrix is symmetric, so we only
+                # need to compute half of it. When X is dense, we can benefit
+                # from the BLAS triangular matrix matrix multiplication `syrk`.
+                distances = _euclidean_dense_dense_fast_sym(X, XX)
+            else:
+                distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)
+                _add_norms(distances, XX, YY)
+
+    # For n_features <= 32, we use the 'exact' method, i.e. the usual method,
+    # d(x,y)² = ||x - y||².
     else:
-        YY = row_norms(Y, squared=True)[np.newaxis, :]
 
-    distances = safe_sparse_dot(X, Y.T, dense_output=True)
-    distances *= -2
-    distances += XX
-    distances += YY
-    np.maximum(distances, 0, out=distances)
+        # distances being between rows of X and Y, it's more efficient to work
+        # on C-contiguous arrays
+        if not issparse(X):
+            X = np.asarray(X, order='C')
+        if not issparse(Y):
+            Y = np.asarray(Y, order='C')
+
+        # Euclidean distance between 2 sparse vectors is very slow. It's much
+        # faster to densify one. We densify the smaller one for lower memory
+        # usage.
+        if issparse(X) and issparse(Y):
+            if Y.shape[0] > X.shape[0]:
+                X = X.toarray()
+            else:
+                Y = Y.toarray()
+
+        if issparse(X):
+            distances = _euclidean_sparse_dense_exact(
+                X.data, X.indices, X.indptr, Y, YY)
+        elif issparse(Y):
+            distances = _euclidean_sparse_dense_exact(
+                Y.data, Y.indices, Y.indptr, X, XX).T
+        else:
+            distances = _euclidean_dense_dense_exact(X, Y)
 
+    # Ensure that distances between vectors and themselves are set to 0.0.
+    # This may not be the case due to floating point rounding errors.
     if X is Y:
-        # Ensure that distances between vectors and themselves are set to 0.0.
-        # This may not be the case due to floating point rounding errors.
-        distances.flat[::distances.shape[0] + 1] = 0.0
+        np.fill_diagonal(distances, 0)
 
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+def _check_norms(X, Y=None, X_norm_squared=None, Y_norm_squared=None):
+    n_features = X.shape[1]
+
+    if n_features > 32 and X.dtype == np.float32:
+        # In this case, we compute euclidean distances by upcasting to float64.
+        # It' necessary to compute the norms on upcasted X and not to upcast
+        # the norms computed on X to keep good precision.
+        return None, None
```

Comment:
Again, testing only the type of `X`.
Moreover, at this point, we explicitely discard the squared norms if they were passed in. I guess this would at least need to be documented in `euclidean_distances`.
And maybe they should only be discarded is they're in float32? (Just an idea.)

Reply:
+1 for documenting that the norms may be ignored in some cases :)

Reply:
Norms are ignored only in the case: `n_features > 32 and dtype=float32` (for float64 no need to upcast so no need to chunk). In that case I can't use norms computed on float32 data. This is the main reason of the loss of precision. So I need to first upcast X and then compute the norms.

I agree I should add a not about that in the docstring.

Reply:
What if the norm vectors are float64 but matrices are float32?
KMeans and birch clusterings rely on these arguments. So if their matrices contain float32, the norm vector wouldn't be used at all and they would have pre-computed them for nothing.

Actually, I think it's a bit weird to expose to the user of the function such a detail of implementation. The user of `euclidean_distances` shouldn't have to know whether it uses the formula `||x||² - 2x.y + ||y||²` or any other. But that's beyond the discussion of this PR.

Reply:
It seems not to happen in sklearn otherwise tests would fail.

But I agree with you, I've been too hard on enforcing recomputing the norms. It's a bit tricky however because if X is float32 and norms are float32 then de have to recompute them.

Reply:
It seems not to happen in sklearn otherwise tests would fail.

But I agree with you, I've been too hard on enforcing recomputing the norms. It's a bit tricky however because if X is float32 and norms are float32 then de have to recompute them.

Reply:
It seems not to happen in sklearn otherwise tests would fail.

But I agree with you, I've been too hard on enforcing recomputing the norms. It's a bit tricky however because if X is float32 and norms are float32 then de have to recompute them.

Reply:
When I say they rely on it, I mean they explicitely precompute the norm vectors in hope of accelerating the complutation of the euclidean distances. see `sklearn/cluster/k_means_.py and `sklearn/cluster/birch.py`. If we don't use it, then it's just useless work they do.

Overall I don't think it should be up to the caller to precompute these vectors as they might not be used anyway when using the exact implementation. They should probably be an optional result of the `euclidean_distances` function.
But this might deserve an issue of its own to find the best strategy.

Reply:
> When I say they rely on it, I mean they explicitely precompute the norm vectors in hope of accelerating the complutation of the euclidean distances. see sklearn/cluster/k_means_.py and sklearn/cluster/birch.py`. If we don't use it, then it's just useless work they do.

I know but if norms are precomputed in float32 then they are useless, the precision is lost and upcasting does not bring it back. They need to be precomputed in float64.

Reply:
Maybe a `dtype` argument could be added to `row_norms` and KMeans and Birch could use it. But that would be for another PR, I guess.

With regards to this PR, maybe discard the norm vectors if their type is float32. That's, after all, the only case where we either won't need it or will need to recompute it.

To simplify the code a bit, maybe start with `XX = None`, keep `X_norm_squared` if its type is float64, and compute it only if `X.dtype` is float64 and `n_features > threshold`.
And since `XX` and `YY` might only be used when `n_features > threshold`, maybe call `_check_norms` only in this case so that its logic can be simplified further.

Reply:
> Maybe a dtype argument could be added to row_norms and KMeans and Birch could use it. But that would be for another PR, I guess.

I agree

> With regards to this PR, maybe discard the norm vectors if their type is float32. That's, after all, the only case where we either won't need it or will need to recompute it.

We need them when input is sparse